### PR TITLE
Add cudnn_rnn_flatten_weight meta kernel

### LINKED ISSUE #120626: [export] `nn.GRU` fails to `torch.export` due to unimplemented operator
### 🐛 Describe the bug

The following model, containing a single `nn.GRU` layer cannot be exported:
 
**Model:**
```python
import torch
from torch.export import export

class Sample(torch.nn.Module):
   def __init__(self, size, h_dim, **kwargs):
       super(Sample, self).__init__(**kwargs)
       self.rnn = torch.nn.GRU(size, h_dim, batch_first=True)
 
   def forward(self, x):
       _, states = self.rnn(x)
       return states

mod = Sample(size=19, h_dim=40)
mod.eval()
mod.cuda()
exported = export(mod, (torch.rand((1, 40, 19)).cuda(),))
```

**Error Log:**
```bash
  File "/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1713, in run_fallback_kernel
    return pytree.tree_map(map_out, r)
  File "/python3.10/site-packages/torch/utils/_pytree.py", line 898, in tree_map
    return treespec.unflatten(map(func, *flat_args))
  File "/python3.10/site-packages/torch/utils/_pytree.py", line 734, in unflatten
    leaves = list(leaves)
  File "/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1703, in map_out
    raise orig_not_implemented_exception
  File "/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1436, in _dispatch_impl
    r = func(*args, **kwargs)
  File "/python3.10/site-packages/torch/_ops.py", line 571, in __call__
    return self_._op(*args, **kwargs)
NotImplementedError: aten::_cudnn_rnn_flatten_weight: attempted to run this operator with Meta tensors, but there was no abstract impl or Meta kernel registered. You may have run into this message while using an operator with PT2 compilation APIs (torch.compile/torch.export); in order to use this operator with those APIs you'll need to add an abstract impl.Please see the following doc for next steps: https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit
```

### Versions

```
torch == 2.3.0.dev20240222+cu121
```

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by angelayi**
I'm not sure if this is a correct meta kernel 😅 
Fixes https://github.com/pytorch/pytorch/issues/120626

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -5784,6 +5784,26 @@ def _cudnn_rnn(
...
+    num_layers,
+    batch_first,
+    bidirectional,
+):
+    torch._check(len(weight_arr) > 0)
+    num_weights = sum(weight.numel() for weight in weight_arr)
+    weight_buf = weight_arr[0].new_empty(
+        num_weights, requires_grad=weight_arr[0].requires_grad
+    )
+    return weight_buf
```
Comment (eellison) [inline torch/_meta_registrations.py:5804]: I believe this `_cudnn_rnn_flatten_weight` does in-place stride modifications to the input parameters. See, this comment here: https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L203-L204

We would to in-place update the metas of the inputs.

**Review by eellison — COMMENTED** [review]
I think this kernel is metdata-mutating. Adding support for it and rnn might be a little tricky, and require a larger change in approach. 

What is the motivation here ?

**Review by ezyang — APPROVED** [review]
meta is not accurate. Maybe a stopgap is to test if the parameters are already in the correct stride (and error if they are not).

**Review by ezyang — CHANGES_REQUESTED** [review]
oops clicked wrong button, see above

**Pull request closed without merging**